[inductor] Allow specify a subdir to store .so and .cubin files

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105466

Summary: The subdir is used to store .so and .cubin files generated by AOTInductor. It can either be specified, or created based on hash of the input graph.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

Differential Revision: [D47556730](https://our.internmc.facebook.com/intern/diff/D47556730)

desertfire committed:
[inductor] Allow specify a subdir to store .so and .cubin files

Summary: The subdir is used to store .so and .cubin files generated by AOTInductor. It can either be specified, or created based on hash of the input graph.

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47556730).

chenyang78 commented on torch/_inductor/codecache.py:
```diff
@@ -299,8 +299,14 @@ def code_hash(code, extra=""):
+        else:
+            subdir = os.path.join(cache_dir(), specified_dir)
+    else:
+        subdir = os.path.join(cache_dir(), basename[1:3])
```
Our of curiosity, why do we select ```[1:3]``` instead of ```[0:2]```?

desertfire replied:
Random historical choice, AFAIK

chenyang78 commented on torch/_inductor/codecache.py:
```diff
@@ -314,10 +320,14 @@ def get_hash(content: Union[str, bytes], extra="", hash_type="code"):
-    content: Union[str, bytes], extension: str, extra="", hash_type: str = "code"
+    content: Union[str, bytes],
+    extension: str,
+    extra="",
```
nit - add type hint to ```extra```?

chenyang78 commented on torch/_inductor/codecache.py:
```diff
@@ -813,20 +828,19 @@ def compile(cls, graph, source_code, cuda):
-                    os.makedirs(output_so_dir, exist_ok=False)
-                so_name = f"{config.dll_name}.so"
-                output_so = os.path.join(output_so_dir, so_name)
+                output_so = f"{input_path[:-4]}.so"
```
nit - could we use ```os.path.splitext``` to remove the file suffix instead of hard-coding ```[:-4]```, which has an assumption about the file ext.

chenyang78 commented on torch/_inductor/config.py:
```diff
@@ -54,8 +54,11 @@
+# If an absolute path is specified, the generated lib files will be stored under the directory;
+# If a relative path is specified, it will be used as a subdirectory under the default caching path;
+# If not specified, a temp directory will be created under the default caching path
+aot_inductor_output_path = ""
```
I think the dll_name will be part of the ```aot_inductor_output_path```?

chenyang78 reviewed (approved):
Left some minor comments. Otherwise, LGTM. Thanks!

desertfire committed:
Update on "[inductor] Allow specify a subdir to store .so and .cubin files"


Summary: The subdir is used to store .so and .cubin files generated by AOTInductor. It can either be specified, or created based on hash of the input graph.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D47556730](https://our.internmc.facebook.com/intern/diff/D47556730)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47556730).

desertfire commented:
@pytorchbot merge
